pkg/http/handlers/base.go

```diff
@@ -0,0 +1,121 @@
+package handlers
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"io"
+	"net/http"
+
+	cerrors "github.com/contiamo/go-base/pkg/errors"
+	"github.com/contiamo/go-base/pkg/tracing"
+	"github.com/pkg/errors"
+)
+
+const (
+	// Megabyte is a pre-defined maximum payload size that can be used in
+	// NewBaseHandler
+	Megabyte = 1024 * 1024
+)
+
+// BaseHandler contains all the base functions every handler should have
+type BaseHandler interface {
+	tracing.Tracer
+	// Read reads the body and tries to decode JSON from it to the given output type
+	Parse(r *http.Request, out interface{}) error
+	// Write writes the response serving the given object as JSON with the given status
+	Write(ctx context.Context, w http.ResponseWriter, status int, obj interface{})
+	// Error serves the proper error object based on the given error and its type
+	Error(context.Context, http.ResponseWriter, error)
+}
+
+// NewBaseHandler creates a new base HTTP handler that
+// contains shared logic among all the handlers.
+// The handler supports parsing and writing JSON objects
+// `maxBodyBytes` is the maximal request body size
+// `componentName` is used for tracing to identify to which
+// component this handler belongs to.
+func NewBaseHandler(componentName string, maxBodyBytes int64, debug bool) BaseHandler {
```

Comment:
Are we going to allow accidental bad values, like 0 or negative values?

Reply:
Well, we propagate it to `LimitReader` that does not check it either. I think, since it's not a user input, it's fine.

Reply:
Added the default `Megabyte` if it's < 0.